### ISSUE TITLE
Loading from local file system not supported.

### ISSUE THREAD

**Issue by acl0056** [mudcube/MIDI.js#74]
DOMLoader does not work for local file system.  In my case, I was able to do a quick fix using jQuery, which I am posting in case it might work for anyone else.

```
DOMLoader.sendRequest = function(conf) {
    var req = new XMLHttpRequest();
    req.open(conf.data ? "POST" : "GET", conf.url, true);
    if (req.overrideMimeType) req.overrideMimeType("text/plain; charset=x-user-defined");
    if (conf.data) req.setRequestHeader('Content-type','application/x-www-form-urlencoded');
    if (conf.responseType) req.responseType = conf.responseType;
    if (conf.onerror) req.onerror = conf.onerror;
    if (conf.onprogress) req.onprogress = conf.onprogress;
    req.onreadystatechange = function (event) {
        if (req.readyState === 4) {
            if (req.status === 0 && jQuery && conf.onload) {
                jQuery.get(conf.url, function(data, textStatus, jqXHR){
                    conf.onload(jqXHR);
                });
                return;
            }
            if (req.status !== 200 && req.status != 304) {
                if (conf.onerror) conf.onerror(event, false);
                return;
            }
            if (conf.onload) {
                conf.onload(req);
            }
        }
    };
    req.send(conf.data);
    return req;
};
```

**Comment by QuinDennis** [comment]
Greetings, I'm working on the same functionality, trying to work with local files only (MIDI.js, jquery, acoustic_grand_piano-ogg.js, local .mid file), then moving it into a phonegap app.  Are there updates?  I'm getting a CORs error on the ogg.  I've even included the ogg.js file, so shouldn't even need to load the soundfont.

**Comment by acl0056** [comment]
If you are not in PhoneGap yet and using Chrome, launch Chrome from the command line with this flag:

--allow-file-access-from-files

Even easier, in Safari with developer mode turned on, check Disable Local File Restrictions under the Develop menu.

**Comment by QuinDennis** [comment]
Wow, thanks for the quick response.

C:\Users\Quin>"C:\Program Files 
(x86)\Google\Chrome\Application\chrome.exe" --allow-file-access-from-files

--and this error in the modified XMLDomLoader---
XMLHttpRequest cannot load 
file:///C:/PW/MIDI/MIDI.js-master/soundfont/acoustic_grand_piano-ogg.js. 
Cross origin requests are only supported for protocol schemes: http, 
data, chrome, chrome-extension, https, chrome-extension-resource.
DOMLoader.XMLHttp.js:136 Uncaught NetworkError: Failed to execute 'send' 
on 'XMLHttpRequest': Failed to load 
'file:///C:/PW/MIDI/MIDI.js-master/soundfont/acoustic_grand_piano-ogg.js'.

**Comment by acl0056** [comment]
You may have to make sure the file you are loading is with the site root as well, meaning the directory that your HTML is in or a subdirectory.

**Comment by QuinDennis** [comment]
Thanks, it is.  mtst.html is in MIDI.js-master



---mtst.html---
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" 
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns = "http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="content-type" content="text/html; charset=utf-8" />
  
  <script src="./js/MIDI/AudioDetect.js" type="text/javascript"></script>
  <script src="./js/MIDI/LoadPlugin.js" type="text/javascript"></script>
  <script src="./js/MIDI/Plugin.js" type="text/javascript"></script>
  <script src="./js/MIDI/Player.js" type="text/javascript"></script>
  <script src="./js/Window/DOMLoader.XMLHttp.js" 
type="text/javascript"></script>
  <script src="./js/jquery-1.11.2.min.js" type="text/javascript"></script>
  <script src="./js/cors.js" type="text/javascript"></script>
  
  <script src="./inc/Base64.js" type="text/javascript"></script>
  <script src="./inc/base64binary.js" type="text/javascript"></script>
  <script src="acoustic_grand_piano-ogg.js" type="text/javascript"></script>
  <script src="play.js" type="text/javascript"></script>
</head>
<body>
<script type="text/javascript">

window.onload = function () { play(); }

</script>
</body>
</html>

-----------------------
//Play
function play() {
     //alert(MIDI.get_audio_status());
     //MIDI.Player.loadMidiFile('hinematov.mid');
     MIDI.loadPlugin({
         soundfontUrl: "file://C:/PW/MIDI/MIDI.js-master/soundfont/",
         instrument: "acoustic_grand_piano",
         callback: function() {
             var delay = 0; // play one note every quarter second
             var note = 50; // the MIDI note
             var velocity = 127; // how hard the note hits
             // play the note
             MIDI.setVolume(0, 127);
             MIDI.noteOn(0, note, velocity, delay);
             MIDI.noteOff(0, note, delay + 0.75);
         }
     });
};

----------------------------

**Comment by acl0056** [comment]
Hmmm... In my project on my Mac, that error goes away when I launch Chrome with that flag.  And I am using file://.  If that is a dead end, I might recommend setting up a local server so you can load through http.  If you don't know what server to use, I might suggest node.js with express.

**Comment by QuinDennis** [comment]
Yep, that should work.  The end game is to get this into a phone app, so 
we'd like to use local files.  I think I'm going to have to look at 
modifying MIDI.js to use FileSystemObject.  I'll see if anyone else is 
interested.

Thanks for your help and quick responses.

And yes, gotta love that node.js!

**Comment by obaidalikhan** [comment]
local system is no showing output

**Comment by acosme** [comment]
@QuinDennis do you achieve "modifying MIDI.js to use FileSystemObject"?
i am going to try it with an Ionic App...

**Comment by Alexis-B** [comment]
If this can be of any interest, I could read local file with firefox without problem using ObjectUrl:

```
<input type="file"
       onchange="handleLocalFile(this.files)" accept="audio/mid"/>
```

```
function handleLocalFile(files) {
	if(files.length != 1){
		console.log('Invalid file selection');
		return;
	}
	f = files[0];
	if(f.type!='audio/mid')
	{
		console.log('Unable to load file');
		return;
	}

	var objectURL = window.URL.createObjectURL(f);

	midi = MIDI.Player;
	midi.loadFile( objectURL , midi.start);
}
```

**Comment by acosme** [comment]
Using ionic + capacitor to run a .mid file:

`   var path = this.file.applicationDirectory + "public/assets/sound/quartet/1.mid";

    window.resolveLocalFileSystemURL(path, (fileEntry) => {

      fileEntry.file((resFile) => {
        var reader = new FileReader(); //const reader = (fileReader as any)["__zone_symbol__originalInstance"];
        reader.onloadend = (evt: any) => {

          MIDI.player.loadFile(evt.target.result, ...)
        };
        reader.onerror = (e) => {
          console.log("Failed file read: " + e.toString());
        };
        reader.readAsDataURL(resFile);

      });
    }, (err) => {

    });`
